File: College labs/AOA in c/sorts/mergeSort.cpp
#include<bits/stdc++.h>
using namespace std;

void Merge(int arr[], int left, int mid, int right){
    int temp[right-left+1];
    int i=left,j=mid+1,k=0;

    while(i<= mid && j<= right){
        if(arr[i]<arr[j]){
            temp[k] = arr[i];
            i++;
        }
        else{
            temp[k] = arr[j];
            j++;
        }
        k++;
    }
    while(i<=mid){
        temp[k] = arr[i];
        i++;
        k++;
    }
    while (j<=right){
        temp[k]= arr[j];
        j++;
        k++;
    }

    for(i=0;i<right-left+1;i++){
        arr[i+left] = temp[i];
    }
    
}

void mergesort(int arr[], int left, int right){
    if(left < right){
        int mid = (right + left)/2;
        // cout<<mid<<" ";
        mergesort(arr, left, mid);
        mergesort(arr, mid+1, right);
        Merge(arr, left, mid, right);
    }
}

int main(){
	//vector <int> arr{4,2,5,6,1,3,7};// 1,3,5,|2
    int n= 7;
	int arr[n] = {4,2,5,6,1,3,7};
    
    mergesort(arr, 0, n-1);

    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
	return 0;
}

File: College labs/AOA/Expt2/MergeSort.java
import java.util.Scanner;

public class MergeSort {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        System.out.print("Enter the number of elements : ");
        int n = sc.nextInt();
        int[] arr = new int[n];

        System.out.print("Enter the Elements : ");
        for(int i=0;i<n;i++) arr[i]= sc.nextInt();
        System.out.println("Passes :");
        mergeSort(arr, 0, arr.length-1);
        System.out.println("\nSorted array :");
        for (int k=0;k<n;k++) {
            System.out.print("\t"+arr[k]);
        }
        sc.close();
    }

    private static void mergeSort(int[] arr, int lb, int ub) {
        if (lb < ub) {
            for (int l = lb; l <= ub; l++)
                System.out.print(arr[l] + " ");
            System.out.println();

            int mid = (lb + ub) / 2;
            mergeSort(arr, lb, mid);
            mergeSort(arr, mid + 1, ub);
            merge(arr, lb, mid, ub);

        }
    }

    private static void merge(int arr[], int lb, int mid, int ub) {
        int i = lb, j = mid + 1, k = 0;
        int[] tempArr = new int[ub-lb+1];
        while (i <= mid && j <= ub) {
            if (arr[i] < arr[j]) {
                tempArr[k] = arr[i];
                i++;
            } else {
                tempArr[k] = arr[j];
                j++;
            }
            // if(k>=(ub-lb+1)) System.out.println(k+":k ub:"+ub+" lb:"+lb);
            k++;
        }
        if (i > mid) {
            while (j <= ub) {
                tempArr[k] = arr[j];
                j++;
                k++;
            }
        } else {
            while (i <= mid) {
                tempArr[k] = arr[i];
                i++;
                k++;
            }
        }

        for (k = lb; k <= ub; k++) { //copy tempArr to main arr
            arr[k] = tempArr[k-lb];
        }
    }
}